contracts/modules/capital/SwapOperator.sol

```diff
@@ -86,129 +78,258 @@ contract SwapOperator {
+  function validateBuyTokenBalance(
+    IPool pool,
+    GPv2Order.Data calldata order,
+    SwapDetails memory buySwapDetails
+  ) internal view {
+    uint buyTokenBalance = order.buyToken.balanceOf(address(pool));
+    
+    // skip validation for WETH since it does not have set swapDetails
+    if (address(order.buyToken) == address(weth)) {
+      return;
+    }
 
-      // Ask oracle how much of the other asset we should get
-      uint oracleBuyAmount = priceFeedOracle.getAssetForEth(address(order.buyToken), order.sellAmount);
+    if (buyTokenBalance >= buySwapDetails.minAmount) {
+      revert InvalidBalance(buyTokenBalance, buySwapDetails.minAmount, 'min');
+    }
+    // NOTE: use order.buyAmount to get postBuyTokenSwapBalance
+    uint postBuyTokenSwapBalance = buyTokenBalance + order.buyAmount; 
+    if (postBuyTokenSwapBalance > buySwapDetails.maxAmount) {
+      revert InvalidPostSwapBalance(postBuyTokenSwapBalance, buySwapDetails.maxAmount, 'max');
+    }
+  }
 
-      // Calculate slippage and minimum amount we should accept
-      uint maxSlippageAmount = (oracleBuyAmount * swapDetails.maxSlippageRatio) / MAX_SLIPPAGE_DENOMINATOR;
-      uint minBuyAmountOnMaxSlippage = oracleBuyAmount - maxSlippageAmount;
+  /// @dev Helper function to determine the SwapOperationType of the order
+  /// NOTE: ETH orders has WETH address because ETH will be eventually converted to WETH to do the swap
+  function getSwapOperationType(GPv2Order.Data memory order) internal view returns (SwapOperationType) {
+    if (address(order.sellToken) == address(weth)) {
+      return SwapOperationType.EthToAsset;
+    }
+    if (address(order.buyToken) == address(weth)) {
+      return SwapOperationType.AssetToEth;
+    }
+    return SwapOperationType.AssetToAsset;
+  }
 
-      require(order.buyAmount >= minBuyAmountOnMaxSlippage, "SwapOp: order.buyAmount too low (oracle)");
+  /// @dev Performs pre-swap validation checks for the given order
+  function performPreSwapValidations(
+    IPool pool,
+    GPv2Order.Data calldata order,
+    SwapOperationType swapOperationType,
+    uint totalOutAmount
+  ) internal view {
+    // NOTE: for assets that does not have any set swapDetails such as WETH it will have SwapDetails(0,0,0,0)
+    SwapDetails memory sellSwapDetails = pool.getAssetSwapDetails(address(order.sellToken));
+    SwapDetails memory buySwapDetails = pool.getAssetSwapDetails(address(order.buyToken));
+
+    // validate both sell and buy tokens are enabled
+    validateTokenIsEnabled(address(order.sellToken), sellSwapDetails);
+    validateTokenIsEnabled(address(order.buyToken), buySwapDetails);
+
+    // validate ETH balance is within ETH reserves after the swap
+    if (swapOperationType == SwapOperationType.EthToAsset) {
+      uint ethPostSwap = address(pool).balance - totalOutAmount;
+      if (ethPostSwap < minPoolEth) {
+        revert EthReserveBelowMin(ethPostSwap, minPoolEth);
+      }
+    }
```

Comment:
This check can be moved inside `validateSellTokenBalance` function where we validate the min/max.